### ISSUE TITLE
Sex - curation before uploading first vocabulary version

### ISSUE THREAD

**Issue by ManonGros** [gbif/vocabulary#83]
Here is a file to edit: https://drive.google.com/file/d/1qyBLQnpLyF3qXlNJ2oP0QN6uoHbIbhxd/view?usp=sharing

**It contains:**

- the list of existing concepts
- a list of the values already mapped to the concepts (they are all in the **Hidden** sheet/tab for now)
- the GBIF verbatim values for this field that appear more than 10,000 times or in 5 or more datasets

NB: We can make sure that the matching doesn't take into account any numbers for the values. For example, no need to match `1 Macho | 1 Hembra` and `1 Macho | 2 Hembra`, the data can simply be `Macho | Hembra`. We just need to notify @marcos-lg before he does the import.

Pease check instructions here: https://github.com/gbif/vocabulary/issues/70

**Comment by pzermoglio** [comment]
I'd like to work on this one, have requested access.
Thanks

**Comment by timrobertson100** [comment]
As we define this vocabulary, Isabel (DanBIF) writes:

>I have a nice dataset with bees in GBIF https://www.gbif.org/dataset/0528c11e-4074-43bb-829f-13b161bd0a56 .
>I notice that the sex “worker” is excluded in records, see e.g.
>https://www.gbif.org/occurrence/1587094798
>Is there a way to change that, so that Worker is accepted in the controlled vocabulary?

**Comment by pzermoglio** [comment]
Nice dataset indeed :)

But to the question, I'd say no.
"worker" is not a value for sex, it's a caste. No matter how obvious it may be for some areas of expertise that all(?) hymenopteran workers are females, terms should not be used to express concepts that are not contemplated in the definition of such terms. If all workers are females, then it'd be easy enough to fill in the sex field with "female" programatically.

I've seen similar cases happen with sex / reproductive condition / lifestage. 
E.g., "pregnant", for a vertebrate. We humans -biologists, experts in blabla, etc.- (note that there is some knowledge required anyway to make the inferences), can deduce that the animal in question is an adult female. However, "pregnant" is a reproductive condition, it is clearly not a sex nor a lifestage.

Back to castes, I understand that there is no specific field to capture "worker" right now, and that we don't want to loose information.  However, misusing another term (like sex) does not help solve the problem. An alternative would be to try to find community support for creating a new "caste" term in Darwin Core. Otherwise, if just using Simple Darwin Core, "worker" would be best shared under dwc:dynamicProperties, encoded as key:value. (and yes, sadly a bit obscured...)

**Comment by CecSve** [comment]
- [ ] Add English labels to the concepts.
- [ ] Check the already mapped values in the Hidden sheet/tab:
        Correct any errors.
        When needed, move the mapped values from the Hidden sheet to the Concept sheet as an alternative label but do not add        any new concept.
- [ ] Map as many verbatim values as possible and add them to the Concepts and Hidden sheet/tabs:
- [ ] Incorporate VertNet mapping to the Concepts and Hidden sheet/tabs if possible: https://github.com/VertNet/DwCVocabs/blob/master/vocabs/
- [ ] Incorporate the ALA mapping to the Concepts and Hidden sheet/tabs if possible.

**Comment by CecSve** [comment]
I have removed mapping of a sex + ? to the sex, as I would interpret this as unknown/best guess, and it is now mapped to unknown instead.

**Comment by DanBIF** [comment]
W.r.t. comment https://github.com/gbif/vocabulary/issues/83#issuecomment-1001640729 The argumantation in the comment is valid, but I really think it is a pity too loose information on caste for datasets on Hymenoptera. How about "redefining" the term "Sex" to be "Sex/Caste"? - then we could include "worker" as a valid value

**Comment by CecSve** [comment]
> W.r.t. comment [#83 (comment)](https://github.com/gbif/vocabulary/issues/83#issuecomment-1001640729) The argumantation in the comment is valid, but I really think it is a pity too loose information on caste for datasets on Hymenoptera. How about "redefining" the term "Sex" to be "Sex/Caste"? - then we could include "worker" as a valid value

We would want to keep the vocabularies strict to only account for one specific term to avoid confusion for publishers as well as users. Values that refer to e.g. 'queen', 'worker', 'drone' etc. are present in the 'Sex' field in 0.34% of the datasets and in 0.05% of occurrences, so it is not a prevalent issue. It would still be possible to get the values from the verbatim data if you carry out a full download. 

Caste systems/eusociality and other sorts of hierarchical social structures extend beyond Hymenoptera and Insecta, and it might be interesting to try to capture the variation in a DwC term (although for practical purposes it may be too complex and specific to capture in a standard), For now, I will map queen = female, but workers are not necessary females (e.g. in termites) so they will not be interpreted as female (worker would still be in the data as the verbatim sex). I am not aware of whether the term drones is used beyond Hymenoptera(?) but will leave it unmapped for sex, to avoid any misinterpretation.

**Comment by CecSve** [comment]
> NB: We can make sure that the matching doesn't take into account any numbers for the values. For example, no need to match `1 Macho | 1 Hembra` and `1 Macho | 2 Hembra`, the data can simply be `Macho | Hembra`. We just need to notify @marcos-lg before he does the import.

I missed this and have accounted for all the numbers after all... they are the reason for the high number of hidden values.

**Comment by CecSve** [comment]
The vocabulary is ready for you to check now @jhnwllr. Please use the `Hidden_OpenRefine` tab to check - you should compare the `Hidden label` to the `Concept_dropdown`.

**Comment by jhnwllr** [comment]
@CecSve this is my review. Feel free to disagree or question my comments. 

might mean MALE https://en.wikipedia.org/wiki/XO_sex-determination_system
`O`	

I think all of these should be Unknown
```
O"[=F]		
t		
upside down F symbol		
```

Why is this marked Unknown? 
`(Female)`

Seems fair to call this Female
`[size Female]`

This might mean Male https://en.wikipedia.org/wiki/Hemipenis
`1 hemipene, throat white, scales`

This should be Unknown 
`1 juvenile`

I bet this should be Female. 
`1 OV.F`
`1OV. F`
`2 OV. F`
`2 O.F`

(later these are all marked as Female) 
```
OV F
OV. F
OV.F
OVF
OVIG F.
OVIG.F
```

Should be Unknown. 
`1juv.`

Should be Unknown. 
`2?`

Previously such combinations were Male. 
`5M1J`

Should be Unknown. 
`A`

I think this means Adult Male. 
`A M`

Should be Female.
`Adult [Female]`

Should be Unknown. 
`adult hooks`

All Exx ex ex. ect have to be some system... I think this is Male. 
`Ex`

This should be Mixed. 
`F (M juv.)`

Interesting that this one is considered Female but other similar are marked Unknown. 
`F [M listed on tag]`

Should be Mixed or Unknown 
`Femalemale`

I bet this stands for female gonads or something. 
`FG`

Should Hermaphrodite
`hermafrodita`

There are a lot of the bracketed reclassifications. I think we need some standard way of dealing with them. 
`M [cataloged as F]` (marked Male)
`M [F]!` (marked unknown) 

Others with question mark have been marked as unknown, but these are marked as male. 
`M by call [?]`
`M imm.?`
`M juv.?`
`M S.Adult?`
`M, juvenile?`
`M/adult (?)`

I think this should be marked Unknown
`male (?)/female`

Should be mixed
`MaleÂ female`

These are all marked as Unknown but in other places similar are marked as Male or Female. 

```
Prob. F		
Prob. Female		
probable adult female		
Probably [Male]		
probably F		
PROBABLY FEMALE		
PROBABLY MALE		
prop. F.		
```

Should be Female. 
`una hembra`

**Comment by tucotuco** [comment]
In my experience the parenthetical and bracketed entries often signify uncertainty. Whether uncertainty should result in "unknown" or in the suggested probable interpretation should be, at the very least, consistent.

**Comment by CecSve** [comment]
Be aware that and empty Sex field can be populated based on data in dynamicProperties https://github.com/gbif/pipelines/issues/478

**Comment by CecSve** [comment]
Thank you for the thorough breakdown @jhnwllr - I will go through them one by one and based on @tucotuco 's comment:

> In my experience the parenthetical and bracketed entries often signify uncertainty. Whether uncertainty should result in "unknown" or in the suggested probable interpretation should be, at the very least, consistent.

register all uncertainty as `unknown`.

I will let you know if I have any questions.

**Comment by CecSve** [comment]
After consulting with the NAOC work group a while back, I have switched the `unknown` category to the concept `indeterminate` following the BODC vocabulary http://vocab.nerc.ac.uk/collection/S10/current/S105/, since `unknown` implies that no effort was made to determine the sex.

**Comment by CecSve** [comment]
@marcos-lg I am ready to try to import this vocabulary to UAT. However, there are several characters that need to be automatically removed PRIOR to mapping for this to work properly. So I just want to make sure that these characters are removed:
**Remove (trailing):**

```
,
.
'
*
 *
"
 '
 -
-
!
 !
]
)
[
numbers (not zeros)
```

**Remove within text string:**

```
,
/
numbers (not zeros)
;
.
+
|
&
```

**Remove (leading):**

```
numbers (not zeros)
[
,
white space
```

**Recode:**
`[]` and `{}` and `{]` and `[}` to `()`

**Comment by CecSve** [comment]
The vocabulary is now uploaded to PROD: https://registry.gbif.org/vocabulary/Sex

**Comment by CecSve** [comment]
When the vocabulary is ready to be implemented in the pipeline, the following clean-up of the verbatim values should be carried out before mapping the values:

## remove trailing
,
.
'
*
 *
"
 '
white space
dash
!
 !
]
)
[
numbers (not zeros)

## remove within text string
,
/
numbers (not zeros)
;
.
+
|
&

## remove (leading)
numbers (not zeros)
[
,
white space

## recode:
[] and {} and {] and [} to ()

**Comment by marcos-lg** [comment]
I was taking a look at the hidden values that we have and I think we need to redefine the cleanup rules.

We have hidden labels that contain `[` and `]`, for example:

`[FEMALE]`
`--?[=F]`
`?[illeg] [M]`

So if we apply the suggested cleanup they won't match with any concept. For example, a verbatim value `?[illeg] [M]` would be converted to `?(illeg) [M`.

Also, if there are several rules to apply it's important to consider the order. For example, removing leading or trailing `[` might class with the _recode_ rule - in the example above, if I had applied the _recode_ rule first the result would have been `?(illeg) (M)`

@CecSve what are the cases that we want to solve with the cleanup so I can get a better understanding of what we need to clean?

**Comment by CecSve** [comment]
Thank you for checking! I was unsure whether the clean-up made 100% sense (I redid it 3 times). 

I see the two comments I made here are conflicting https://github.com/gbif/vocabulary/issues/83#issuecomment-1956918982 and https://github.com/gbif/vocabulary/issues/83#issuecomment-1957038064. Let us stick to this (probably need to remap some values so the vocabulary is consistent with what users see, though it won't affect interpretation):

1. Anything with a `?` should not be mapped to a concept. `prob.` and `probably` etc. will be treated as `?` and should not be mapped as well. 
2. I think we would want to recode after step 1 and then clean the rest. 
3. () or [] is mapped to the sex supplied. 

I wonder if it is worth me going over it all first with these rules and checking to see if it still makes sense with the clean-up suggested in https://github.com/gbif/vocabulary/issues/83#issuecomment-2202282810? I can't find my script from last time, but then I can share it with you this time?

**Comment by tucotuco** [comment]
@CecSve There are conventions in Mammalogy and probably in other disciplines by association that parentheses signify uncertainty, so strings that contain them should be mapped to indeterminate also.  The square brackets are different, as the convention for that is to signify information that was not recorded originally, but that does not carry with it the uncertainty of the parentheses, so that one should be fine to interpret by dropping the square brackets.

**Comment by CecSve** [comment]
> @CecSve There are conventions in Mammalogy and probably in other disciplines by association that parentheses signify uncertainty, so strings that contain them should be mapped to indeterminate also. The square brackets are different, as the convention for that is to signify information that was not recorded originally, but that does not carry with it the uncertainty of the parentheses, so that one should be fine to interpret by dropping the square brackets.

Thanks for the reminder John! You have mentioned this to me before and I forgot. I will correct the suggestion.

**Comment by marcos-lg** [comment]
It's important to remember that the mappings should be in the labels so the interpretation is transparent and there are no rules in the Java code that can change the mappings of the labels. The only thing we can do is to clean up the values before the interpretation so we don't have to create a hidden label for each possible case. But the cleanup should be for characters that don't have any value. For example, in `LifeStage` we remove the leading numbers because there were a lot of verbatim values like:

- 2 adults
- 3 adults
- 4 adults
- ...

And we can't add a hidden label for every number.

Therefore, I can't do things like this because it's a mapping done in java that doesn't take the vocabulary into account:

> Anything with a ? should be mapped as indeterminate. prob., probably, () etc. will be treated as ? and will also be mapped as indeterminate.

Removing the brackets is fine although I see that we have many hidden labels that contain brackets so it doesn't feel right to me.

**Comment by CecSve** [comment]
> If we are not sure about the need of the cleanup I suggest not to do it and add it a later point if we see it's necessary.

I agree. I have decided to take one last look to see if any cleanup would make sense. So I am remapping everything and will share the JSON with the steps I took.

**Comment by tucotuco** [comment]
@marcos-lg I think I am missing something. If you can clean up by removing things like square brackets, why can you not clean up by substituting "indeterminate", which is in the vocabulary, for anything with the patterns @CecSve mentions?

**Comment by marcos-lg** [comment]
My concern is that the cleanup was intended to remove characters that don't have any significant value but a substitution might override the labels that are present in the vocabulary. 

And I see that we have hidden labels like `?MALE` which makes me think that we have already covered the cases with `?`. In other words, if we do those substitutions these hidden labels will never be used and that doesn't feel right to me.

**Comment by CecSve** [comment]
> And I see that we have hidden labels like `?MALE` which makes me think that we have already covered the cases with `?`. In other words, if we do those substitutions these hidden labels will never be used and that doesn't feel right to me.

To put it in context what I am preparing with the example you gave. 

1. I will map `?M` to `indeterminate`
2. I am cleaning any values that will lead to the same value (`?M`) if:

- some spaces are removed (leading, trailing, whitespace)
- numbers are removed
- special characters are removed (e.g., ',!"# etc.)

I do this to have a shorter list than 3000+ verbatim values to map, leaving me with approximately half the amount of verbatim values to map. I will upload the original sheet and the JSON for the cleaning steps soon.

**Comment by CecSve** [comment]
@marcos-lg I have attached the edits I made in OpenRefine using GREL in JSON format. I haven't attached the full history of what I did since it is only relevant for standardizing concepts and hidden values before vocabulary mapping. We end up with 1.447 verbatim values mapped to 5 concepts.

If you think the following makes sense, then I will update the sex vocabulary based on what I just did:

1. Pipelines will cleanup the verbatim values based on the steps included in this file (maybe a little restructuring is needed, but cleanup result should be the same)
2. Cleaned-up values will get mapped to the vocabulary 
3. The cleanup steps should be documented in the [technical documentation](https://techdocs.gbif.org/en/) (probably in the data processing section)

Does this plan make sense, please?

[value_edits_history.json](https://github.com/user-attachments/files/16816087/value_edits_history.json)

**Comment by tucotuco** [comment]
> My concern is that the cleanup was intended to remove characters that don't have any significant value but a substitution might override the labels that are present in the vocabulary.
> 
> And I see that we have hidden labels like `?MALE` which makes me think that we have already covered the cases with `?`. In other words, if we do those substitutions these hidden labels will never be used and that doesn't feel right to me.

This tells me that the approach is not right. The method isn't sacred, the result is. So what is the goal? I would have thought that the goal is to do the best matching with the least vocabulary maintenance. I know vocabulary maintenance is important, or GBIF would not have decided to discard hidden labels that apply to very few records or very few data sets. @CecSve also mentions it in this issue.

So what happens when "male, maybe, sort of?" starts getting published and isn't in the list of hidden values. The proposed approach means that "male, maybe, sort of" has to be in the hidden values and mapped or it will not be improved. If processing turned everything with a '?' in it into "indeterminate", every appropriate improvement would be made immediately without any reliance on vocabulary maintenance. That seems like a serious win to me.  Again, if I misunderstand something, my apologies.

**Comment by marcos-lg** [comment]
AFAIK the goal was to move the interpretation from java code(enums and parsers) to the vocabulary so it's users who decide how values should be mapped. That's why hardcoding that rule of replacing `?`with `indeterminate` breaks this approach.

One thing we can do is to extend the vocabulary to allow regular expressions (in the hidden labels or as a new field).

**Comment by tucotuco** [comment]
Ok, I understand. It just seemed that if you were doing things in code
(which it still seems like you are), why not do something even more useful?

Regular expressions in the vocabulary would be interesting, but would be
beyond most vocabulary maintenance participants.

if re.search(r'.*[\?\(].*', input_string): return "indeterminate". :-)

**Comment by marcos-lg** [comment]
> Ok, I understand. It just seemed that if you were doing things in code (which it still seems like you are), why not do something even more useful?

Yeah, that's true. To be honest I'm not 100% sure that adding the cleanup in the code was the best decision but it was convenient at that time.

I'll give some thinking to the regular expressions to see if we can allow them without introducing much more problems.

**Comment by marcos-lg** [comment]
Since we have the lookup endpoint in UAT already I did a test with the verbatim values of this vocab(e.g.: https://api.gbif-uat.org/v1/vocabularies/Sex/concepts/lookup?q=macho). I took the verbatim values from prod with this query:

```
SELECT 
  v_sex, count(distinct datasetKey) as d, count(*) as occ 
FROM 
  occurrence
GROUP BY 
  v_sex 
HAVING count(*)>10000 OR count(distinct datasetKey) >=5 
  ORDER BY v_sex
```

I tried the lookup without any cleanup and I got these results:

Matches: 301
Not matches: 1457

Then I played with the suggested cleanups in https://github.com/gbif/vocabulary/issues/83#issuecomment-2320913808 (trimming and the case are not needed since they are handled by the lookup service in the API): 

- rawVerbatim.replaceAll("--", "").replaceAll("[0-9|%|@|#|!|\"|,|*|.|']", "");
Matches: 1623
Not matches: 135


- rawVerbatim.replaceAll("--", "").replaceAll("[0-9|%|@|#|!|\"|,|*|.|'|+|\\|]", "");
Matches: 1623
Not matches: 135


- rawVerbatim.replaceAll("--", "").replaceAll("[0-9|%|@|#|!|\"|,|*|.|'|+|\\|]", "").replaceAll("/", "\\\\");
Matches: 1616
Not matches: 142

So the best cleanup was by removing these characters:

- numbers
- `%`
- `@`
- `#`
- `!`
- `"`
- `,`
- `*`
- `.`
- `'`

Then I took a look at the hidden labels and I noticed that we were missing some combinations for the `mixed` concept so I added hidden labels to the `mixed`concept with all the possible combinatiosn of `male` and `female` in singular and plural and in any order and with the conjunctions `,`, `|` and `&`. And I did the same with `macho`and `hembra`and the conjunctions `,`, `|` and `y`. You can see them here: https://api.gbif-uat.org/v1/vocabularies/Sex/concepts/Mixed/hiddenLabels?limit=48

Having those labels I ran the lookup by using a cleanup that only removes numbers and the results were:

Matches: 1645
Not matches: 113

The verbatims values that weren't mapped were this:

```
"NULL"
"*"
"-"
"--"
"0"
"1"
"1 female, 1 juvenile"
"1 female, 2 juveniles"
"1 female, 4 juveniles"
"1 macho, 1 fêmea"
"1 male, 1 juvenile"
"1 worker"
"10"
"11"
"12"
"13"
"14"
"15"
"16"
"17"
"18"
"19"
"1f,1m"
"1 female"
"2"
"2 Indeterminados"
"2 fêmeas"
"20"
"21"
"22"
"23"
"24"
"27"
"28"
"2f,1m"
"3"
"30"
"31"
"32"
"3f,1m"
"4"
"4 Indeterminados"
"5"
"50"
"6"
"6 Indeterminados"
"7"
"8"
"9"
"B"
"Dioecious"
"Dioico"
"Egg"
"Ex"
"F, M"
"F,M"
"Female;Male;Unknown"
"FÍmea"
"G"
"H"
"HCH"
"HE"
"HS"
"HSH"
"Hc/h"
"I"
"Inconnu"
"Indéterminé"
"L"
"LOT"
"M; F"
"MC"
"Mixte"
"Monoica (flores femeninas y masculinas en un mismo individuo)"
"N"
"ND"
"Non renseigné"
"P"
"Q"
"Staminate"
"Una hembra y un macho"
"Unchecked"
"Worker"
"X"
"_"
"`"
"egg"
"ex"
"exuvia"
"fc"
"female ?; male"
"female; female; female"
"female; female; male"
"female; male; male"
"fêmea (operária)"
"h"
"i"
"i par"
"indeterminate."
"indéterminé"
"l"
"larva/juvenile"
"n"
"nymphe"
"operária"
"ovigerous female"
"pistillate"
"soldier"
"staminate"
"sterile"
"worker"
"workers"
"x"
```

We could add some other hidden labels to fix some of those cases too.

There are also many hidden labels that are not needed because the cleanup takes care of those cases such as:

```
10 Males 10 Females
10 Males Female
10 Males, 2 Females
etc.
```

but we can leave them there since they don't do any harm and we don't spend time on that.

I did this because the less cleanup we do the better since we have less logic in the java code. Because changes in the java code take us more time to do than changes in the labels.

@CecSve @tucotuco if you guys agree with this I'll create the hidden labels in prod and start integrating this vocabulary into the pipelines interpretation with the cleanup that removes numbers only.

**Comment by ArthurChapman** [comment]
As discussed with @CecSve and @ymgan at TDWG2024 in Okinawa, the operation of tests against the GBIF Vocabularies in the proposed BDQ Core work better if we can test against as many as possible terms - especially where these can be linked as synonyms. As mentioned by @marcos-lg (https://github.com/gbif/vocabulary/issues/83#issuecomment-2340850055) above we would strongly support the suggestion of @tucotuco of retaining all published values (except for those cleaned up such as numbers, etc.). This would apply to all the vocabularies maintained by GBIF.

**Comment by CecSve** [comment]
[Arctos code tables](https://arctos.database.museum/info/ctDocumentation.cfm?table=ctsex_cde) have male? and female? concepts, which GBIF translates to `indeterminate`
